### ISSUE TITLE
client.multi() hgetall performance

### ISSUE THREAD

**Issue by bhartsb** [NodeRedis/node_redis#1225]
node version: v7.4.0
node-redis version: v.2.6.2
redis_version:2.8.4
ubuntu 14.04 2GB RAM VPS instance

I'm getting indications that node_redis hgetall performance isn't what it should be, but possibly I'm doing something wrong.   I have a sorted set with 31669 elements.  The elements are hash keys and each hash has 14 fields and is ~256 bytes.  Retrieving the hashes takes ~64 seconds which seems too slow.
Here is what the function looks like:

`function getAllAnnotations() 
    {
        var currentSeconds = Math.floor((new Date()).getTime() / 1000);
        console.log('currentSeconds before zrange: ' + currentSeconds);

        client.zrangebyscore("geoHashSortedSet", "-inf", "+inf", function(err, reply) {
            
            multi = client.multi();
            for (var uuid in reply) {
                multi.hgetall(reply[uuid]);

            }
            multi.exec(function(err, replies) {
                
                var currentSeconds = Math.floor((new Date()).getTime() / 1000);
                console.log('currentSeconds after multi returns: ' + currentSeconds);
                allAnnotations = replies;
            });

        });

    }`

If I'm not doing anything wrong (???) and there is indeed a performance issue, then I can take more time to provide code to generate the hashes and the sorted set with the hash keys.

**Comment by stockholmux** [comment]
First things first, try using `console.time()` and `console.timeEnd()` ([more info](https://nodejs.org/api/console.html#console_console_time_label)) to profile what's going on. 

I would look at a few thing - how long is it taking for the `zrangebyscore` to return back the `reply` and how long it's taking for your `for...in` loop to add 31,669 to the `multi`. Finally, look at how long it's taking for the `multi.exec` to return a reply. That will help you diagnose why it's taking a while.

I suspect a couple of things, but I can't be sure until we understand how long each step is taking.
Here are my off-the-cuff guesses:

- the `for ... in` with the `hgetall` is triggering a garbage collection. 31k is a big array, after all.
- The sheer network transport time of a 31k part `multi`/`exec` is considerable. And, If I'm understanding what you're doing ,you've got a ton of data coming back: 14 fields * 256 bytes * 31,116 = 111,698,944 bytes (or 111.69 mb). Also `hgetall` returns the field names, so you'll need to count those too for total size.

On a non-Node.js related note, you might want to reconsider what you're doing here - a `multi`/`exec` that large is probably really blocking your Redis server.

**Comment by BridgeAR** [comment]
@bhartsb you are very likely using a older redis-parser version that still had issues with big nested arrays. The implementation had a flaw that was overlooked in the beginning.

Your issue is going away as soon as you update your dependencies. As a side note: by using `batch` instead of `multi` your performance should again go up if you do not have the necessity for having a transaction.

Overall your call should not take long with that change.

@stockholmux Redis can easily handle grabbing millions of keys per second. This operation is not as heavy as it seems to you.

**Comment by stockholmux** [comment]
I had a few moments and did a best case scenario setup:
- local Redis instance
- empty DB
- no traffic
- short as possible keys: zset had a key of 'z' and the hash fields were `f0` ... `x13`, etc. Each field was filled with 256 bytes of `a`

I'm running on a very healthy, modern dev machine.

With node_redis, I was still getting 1-1.5 seconds on average to complete the function. Granted, no where near the 64 seconds of the OP (which is probably accounted by the above mentioned improvements on the nested array parser). But still, this operation, as-is, would be a little pokey. So, @bhartsb that's about what you should expect in ideal circumstances.


@BridgeAR It's still heavy, IMHO.

**Comment by bhartsb** [comment]
@stockholmux thanks for running the test!
@BridgeAR thank you for providing the solution!  I updated and my time went to under 2 seconds.

Sorry to both that it took me this long to respond, I was feeling under the weather yesterday.

For anyone else looking at this: I referenced the following when updating dependencies:
http://stackoverflow.com/questions/16073603/how-do-i-update-each-dependency-in-package-json-to-the-latest-version
